github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://viewLog.html?buildId=4317589&tab=buildLog) with [artifacts](https://viewLog.html?buildId=4317589&tab=artifacts#/) on master @ [v0.0.0-20220205134134-9faac799ba03](https://github.com/cockroachdb/pebble/commits/v0.0.0-20220205134134-9faac799ba03):


```
        // INFO: [JOB 9632] flushed 2 memtables to L0 [031067] (776 B), in 0.0s (0.0s total), output rate 5.3 M/s
        // INFO: [JOB 9632] WAL deleted 031064
        // INFO: [JOB 9634] WAL created 031068
        // INFO: write stall ending
        db.SingleDelete("gcsandjd@6", false /* maybeReplaceDelete */) // <nil> #14552
        // INFO: [JOB 9635] flushing 2 memtables to L0
        // INFO: [JOB 9635] flushing: sstable created 031069
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 9635] flushed 2 memtables to L0 [031069] (778 B), in 0.0s (0.0s total), output rate 5.7 M/s
        // INFO: [JOB 9636] WAL created 031070
        // INFO: write stall ending
        db.Delete("nbdqxnibcc@17") // <nil> #14553
        // INFO: [JOB 9635] WAL deleted 031066
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 9637] flushing 2 memtables to L0
        // INFO: [JOB 9637] flushing: sstable created 031071
        // INFO: [JOB 9637] flushed 2 memtables to L0 [031071] (781 B), in 0.0s (0.0s total), output rate 7.2 M/s
        // INFO: [JOB 9639] WAL created 031072
        // INFO: [JOB 9637] WAL deleted 031068
        // INFO: write stall ending
        db.Delete("rhyfxffvz@7") // <nil> #14554
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 9640] flushing 2 memtables to L0
        // INFO: [JOB 9640] flushing: sstable created 031073
        // INFO: [JOB 9640] flushed 2 memtables to L0 [031073] (779 B), in 0.0s (0.0s total), output rate 7.3 M/s
        // INFO: [JOB 9640] WAL deleted 031070
        // INFO: [JOB 9642] WAL created 031074
        // INFO: write stall ending
        db.Delete("njsupzuvq@10") // <nil> #14555
        // INFO: [JOB 9643] flushing 2 memtables to L0
        // INFO: [JOB 9643] flushing: sstable created 031075
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 9643] flushed 2 memtables to L0 [031075] (780 B), in 0.0s (0.0s total), output rate 6.7 M/s
        // INFO: [JOB 9645] WAL created 031076
        // INFO: write stall ending
        db.Merge("jhtntxo@9", "drvxi") // <nil> #14556
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 9643] WAL deleted 031072
        // INFO: [JOB 9646] flushing 2 memtables to L0
        // INFO: [JOB 9646] flushing: sstable created 031077
        // INFO: [JOB 9646] flushed 2 memtables to L0 [031077] (782 B), in 0.0s (0.0s total), output rate 7.5 M/s
        // INFO: [JOB 9647] WAL created 031078
        // INFO: write stall ending
        db.Delete("iwwmuyd@6") // <nil> #14557
        // INFO: [JOB 9646] WAL deleted 031074
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 9648] flushing 2 memtables to L0
        // INFO: [JOB 9648] flushing: sstable created 031079
        // INFO: [JOB 9648] flushed 2 memtables to L0 [031079] (777 B), in 0.0s (0.0s total), output rate 6.0 M/s
    --- FAIL: TestMeta/random-007 (1203.31s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://viewLog.html?buildId=4333885&tab=buildLog) with [artifacts](https://viewLog.html?buildId=4333885&tab=artifacts#meta) on master @ [v0.0.0-20220207223349-db46dab683e7](https://github.com/cockroachdb/pebble/commits/v0.0.0-20220207223349-db46dab683e7):


```
        iter149.Prev("") // [false] <nil> #15562
        iter149.SeekGE("gxjhrrsbgvm@15", "zjdoaldv@6") // [valid,"qhvzdth@17","pvca"] <nil> #15563
        iter150.SeekPrefixGE("ahxvwaavqz@7") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #15564
        iter150.SeekPrefixGE("jlnrx@16") // [false] pebble: SeekPrefixGE supplied with key outside of lower bound #15565
        iter150.Last() // [false] <nil> #15566
        // INFO: [JOB 403] flushing 2 memtables to L0
        // INFO: [JOB 401] WAL created 044768
        db.Merge("bguh@16", "pxxzdgayx") // <nil> #15567
        iter150.SeekLT("aixzfkgodlx@7", "") // [false] <nil> #15568
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 400] WAL deleted 044759
        // INFO: [JOB 400] MANIFEST deleted 044761
        // INFO: [JOB 403] flushing: sstable created 044769
        // INFO: [JOB 403] MANIFEST created 044770
        // INFO: [JOB 403] flushed 2 memtables to L0 [044769] (787 B), in 0.0s (0.0s total), output rate 254 K/s
        // INFO: [JOB 403] WAL deleted 044762
        // INFO: [JOB 405] flushing 2 memtables to L0
        // INFO: [JOB 404] WAL created 044771
        // INFO: write stall ending
        db.Merge("ekkkfcp@20", "ed") // <nil> #15569
        // INFO: [JOB 403] MANIFEST deleted 044764
        // INFO: [JOB 405] flushing: sstable created 044772
        iter150.First() // [false] <nil> #15570
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 405] MANIFEST created 044773
        // INFO: [JOB 405] flushed 2 memtables to L0 [044772] (784 B), in 0.0s (0.0s total), output rate 295 K/s
        // INFO: [JOB 405] WAL deleted 044765
        // INFO: [JOB 407] flushing 2 memtables to L0
        // INFO: [JOB 406] WAL created 044774
        // INFO: write stall ending
        db.Set("uswcb@20", "mis") // <nil> #15571
        iter149.SeekGE("gbrxevstxwa@12", "") // [true,"qhvzdth@17","pvca"] <nil> #15572
        // INFO: [JOB 405] MANIFEST deleted 044767
        // INFO: [JOB 407] flushing: sstable created 044775
        iter149.SeekLT("ebfbdsttiej@19", "") // [false] <nil> #15573
        iter149.SetBounds("kgyzwnrp@15", "opcwlcxpqg@9") // <nil> #15574
        // INFO: [JOB 407] MANIFEST created 044776
        // INFO: [JOB 407] flushed 2 memtables to L0 [044775] (780 B), in 0.0s (0.0s total), output rate 242 K/s
        iter149.SeekLT("opcwlcxpqg@9", "") // [false] <nil> #15575
        iter149.Next("") // [false] <nil> #15576
        // INFO: [JOB 409] flushing 2 memtables to L0
        // INFO: [JOB 407] WAL deleted 044768
        // INFO: [JOB 408] WAL created 044777
        // INFO: [JOB 409] flushing: sstable created 044778
        db.Merge("dzafh@20", "xlcfodfjfekb") // <nil> #15577
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 407] MANIFEST deleted 044770
        // INFO: [JOB 409] MANIFEST created 044779
        // INFO: [JOB 409] flushed 2 memtables to L0 [044778] (779 B), in 0.0s (0.0s total), output rate 291 K/s
    --- FAIL: TestMeta/random-015 (1263.36s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
dev test @com_github_cockroachdb_pebble//internal/metamorphic:metamorphic_test -f TestMeta --stress --stress-args='-p 1' -- --define gotags=bazel,invariants --test_timeout=11000 --test_arg -seed --test_arg 1644387077574601391 --test_arg -ops --test_arg "uniform:5000-25000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://viewLog.html?buildId=4343918&tab=buildLog) with [artifacts](https://viewLog.html?buildId=4343918&tab=artifacts#meta) on master @ [v0.0.0-20220209231437-b2295acd0a15](https://github.com/cockroachdb/pebble/commits/v0.0.0-20220209231437-b2295acd0a15):


```
        // INFO: [JOB 705] MANIFEST created 027528
        // INFO: [JOB 705] flushed 2 memtables to L0 [027293 027294 027295 027296 027297 027298 027299 027300 027301 027302 027303 027304 027305 027306 027307 027308 027309 027310 027311 027312 027313 027314 027315 027316 027317 027318 027319 027320 027321 027322 027323 027324 027325 027326 027327 027328 027329 027330 027331 027332 027333 027334 027335 027336 027337 027338 027339 027340 027341 027342 027343 027344 027345 027346 027347 027348 027349 027350 027351 027352 027353 027354 027355 027356 027357 027358 027359 027360 027361 027362 027363 027364 027365 027366 027367 027368 027369 027370 027371 027372 027373 027374 027375 027376 027377 027378 027379 027380 027381 027382 027383 027384 027385 027386 027387 027388 027389 027390 027391 027392 027393 027394 027395 027396 027397 027398 027399 027400 027401 027402 027403 027404 027405 027406 027407 027408 027409 027410 027411 027412 027413 027414 027415 027416 027417 027418 027419 027420 027421 027422 027423 027424 027425 027426 027427 027428 027429 027430 027431 027432 027433 027434 027435 027436 027437 027438 027439 027440 027441 027442 027443 027444 027445 027446 027447 027448 027449 027450 027451 027452 027453 027454 027455 027456 027457 027458 027460 027461 027462 027463 027464 027465 027466 027467 027468 027469 027470 027471 027472 027473 027474 027475 027476 027477 027478 027479 027480 027481 027482 027483 027484 027485 027486 027487 027488 027489 027490 027491 027492 027493 027494 027495 027496 027497 027498 027499 027500 027501 027502 027503 027504 027505 027506 027507 027508 027509 027510 027511 027512 027513 027514 027515 027516 027517 027518 027519 027520 027521 027522 027523 027524 027525 027526 027527] (195 K), in 1.8s (1.8s total), output rate 111 K/s
        // INFO: [JOB 709] WAL created 027531
        // INFO: write stall ending
        db.Merge("yexpuernf@11", "xm") // <nil> #9601
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 711] flushing 2 memtables to L0
        // INFO: [JOB 711] flushing: sstable created 027533
        // INFO: [JOB 705] WAL deleted 027287
        // INFO: [JOB 705] MANIFEST deleted 027291
        // INFO: [JOB 708] MANIFEST created 027532
        // INFO: [JOB 708] compacted(default) L0 [019906 020157 020613 021083 021335 021672 021673 021933 021934 022205 022206 022520 022521 022943 022944 023055 023056 023267 023268 023452 023543 023544 023545 023646 023647 023648 024245 024246 024247 024616 024617 024618 025005 025006 025007 025129 025130 025131 025291 025292 025293 025817 025818 025819 026139 026140 026141 026142 026219 026220 026221 026222 026967 026968 026969 026970 027237 027238 027240] (49 K) + L4 [025376] (974 B) -> L4 [027529 027530] (1.9 K), in 0.0s (1.0s total), output rate 279 K/s
        // INFO: [JOB 711] MANIFEST created 027534
        // INFO: [JOB 711] flushed 2 memtables to L0 [027533] (782 B), in 0.0s (0.9s total), output rate 338 K/s
        // INFO: [JOB 708] sstable deleted 025817
        // INFO: [JOB 713] compacting(move) L4 [024683] (927 B) + L5 [] (0 B)
        // INFO: [JOB 714] WAL created 027536
        // INFO: write stall ending
        // INFO: [JOB 708] sstable deleted 025818
        db.Set("qktpxs@15", "jtellvgjh") // <nil> #9602
        iter92.Prev("") // [false] <nil> #9603
        // INFO: [JOB 715] flushing 2 memtables to L0
        // INFO: [JOB 715] flushing: sstable created 027537
        // INFO: write stall beginning: memtable count limit reached
        // INFO: [JOB 713] MANIFEST created 027535
        // INFO: [JOB 708] sstable deleted 025819
        // INFO: [JOB 713] compacted(move) L4 [024683] (927 B) + L5 [] (0 B) -> L5 [024683] (927 B), in 0.0s (0.0s total), output rate 115 M/s
        // INFO: [JOB 716] compacting(move) L4 [025358] (1019 B) + L5 [] (0 B)
        // INFO: [JOB 713] WAL deleted 027292
        // INFO: [JOB 708] sstable deleted 026139
        // INFO: [JOB 713] MANIFEST deleted 027528
        // INFO: [JOB 708] sstable deleted 026140
        // INFO: [JOB 713] MANIFEST deleted 027532
        // INFO: [JOB 708] sstable deleted 026141
        // INFO: [JOB 708] sstable deleted 026142
        // INFO: [JOB 708] sstable deleted 026219
        // INFO: [JOB 708] sstable deleted 026220
        // INFO: [JOB 708] sstable deleted 026221
        // INFO: [JOB 708] sstable deleted 026222
        // INFO: [JOB 708] sstable deleted 026967
        // INFO: [JOB 708] sstable deleted 026968
        // INFO: [JOB 708] sstable deleted 026969
        // INFO: [JOB 708] sstable deleted 026970
        // INFO: [JOB 708] sstable deleted 027237
        // INFO: [JOB 708] sstable deleted 027238
        // INFO: [JOB 708] sstable deleted 027240
        // INFO: [JOB 708] MANIFEST deleted 027459
        // INFO: [JOB 715] MANIFEST created 027538
        // INFO: [JOB 715] flushed 2 memtables to L0 [027537] (786 B), in 5.5s (5.6s total), output rate 141 B/s
    --- FAIL: TestMeta/random-020 (1138.71s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
dev test @com_github_cockroachdb_pebble//internal/metamorphic:metamorphic_test -f TestMeta --stress --stress-args='-p 1' -- --define gotags=bazel,invariants --test_timeout=11000 --test_arg -seed --test_arg 1644477674737468071 --test_arg -ops --test_arg "uniform:5000-25000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/internal/cache/clockpro.go
+++ b/internal/cache/clockpro.go
@@ -671,22 +671,16 @@ func newShards(size int64, shards int) *Cache {
 		c := obj.(*Cache)
 		if v := atomic.LoadInt64(&c.refs); v != 0 {
 			c.tr.Lock()
-			fmt.Fprintf(os.Stderr, "pebble: cache (%p) has non-zero reference count: %d\n%s",
-				c, v, strings.Join(c.tr.msgs, "\n"))
+			fmt.Fprintf(os.Stderr,
+				"pebble: cache (%p) has non-zero reference count: %d\n", c, v)
+			if len(c.tr.msgs) > 0 {
+				fmt.Fprintf(os.Stderr, "%s\n", strings.Join(c.tr.msgs, "\n"))
+			}
 			c.tr.Unlock()
 			os.Exit(1)
 		}
 	})
 	return c
-}
-
-func (c *Cache) trace(msg string, refs int64) {
-	if invariants.Enabled {
-		s := fmt.Sprintf("%s: refs=%d\n%s", msg, refs, debug.Stack())
-		c.tr.Lock()
-		c.tr.msgs = append(c.tr.msgs, s)
-		c.tr.Unlock()
-	}
 }
 
 func (c *Cache) getShard(id uint64, fileNum base.FileNum, offset uint64) *shard {

--- a/internal/cache/clockpro_normal.go
+++ b/internal/cache/clockpro_normal.go
@@ -1,0 +1,10 @@
+// Copyright 2022 The LevelDB-Go and Pebble Authors. All rights reserved. Use
+// of this source code is governed by a BSD-style license that can be found in
+// the LICENSE file.
+
+//go:build !tracing
+// +build !tracing
+
+package cache
+
+func (c *Cache) trace(_ string, _ int64) {}

--- a/internal/cache/clockpro_tracing.go
+++ b/internal/cache/clockpro_tracing.go
@@ -1,0 +1,20 @@
+// Copyright 2022 The LevelDB-Go and Pebble Authors. All rights reserved. Use
+// of this source code is governed by a BSD-style license that can be found in
+// the LICENSE file.
+
+//go:build tracing
+// +build tracing
+
+package cache
+
+import (
+	"fmt"
+	"runtime/debug"
+)
+
+func (c *Cache) trace(msg string, refs int64) {
+	s := fmt.Sprintf("%s: refs=%d\n%s", msg, refs, debug.Stack())
+	c.tr.Lock()
+	c.tr.msgs = append(c.tr.msgs, s)
+	c.tr.Unlock()
+}

